File: wxbot/whatsapp-bookkeeping/src/parser.ts
// ============================================================
// Bookkeeping - Message Parser
// Pure regex-based, zero LLM dependency, deterministic
// Supports multiple formats: +100rmb, 100rmb+, rmb+100, etc.
// ============================================================

/**
 * Negative categories: sign is REVERSED.
 * +25rg 5.3 => actually -132.5 (expense)
 * -25rg 5.3 => actually +132.5 (reversal)
 */
export const NEGATIVE_CATEGORIES = new Set([
  "rg", "sp", "gs", "xb", "it", "st", "ft", "mx",
  "gg", "dg", "ae", "lulu", "dk", "ebay", "rb",
  "uber", "xc", "cvs", "ymx", "psn", "chime", "ks", "tt", "nike", "nd", "cash", "ps",
]);

/** Positive categories: sign is kept as-is */
export const POSITIVE_CATEGORIES = new Set(["rmb"]);

/** All known categories */
export const ALL_CATEGORIES = new Set([...NEGATIVE_CATEGORIES, ...POSITIVE_CATEGORIES]);

export interface ParsedTransaction {
  /** Original sign from message: 1 for +, -1 for - */
  inputSign: 1 | -1;
  /** Face value / amount number */
  amount: number;
  /** Category code, lowercase. "rmb" if pure RMB */
  category: string;
  /** Exchange rate (null for rmb) */
  rate: number | null;
  /** Final RMB equivalent with correct sign applied */
  rmbValue: number;
  /** Original raw message text */
  raw: string;
}

// ============================================================
// Multiple regex patterns for different input formats
// ============================================================

// Pattern 1: +100rmb, -50sp 4.8 (sign before amount, optional category after, optional rate after category)
const PATTERN1 = /^([+-])(\d+(?:\.\d+)?)\s*([a-zA-Z]+)?\s*(\d+(?:\.\d+)?)?$/;

// Pattern 2: 100rmb+, 25rg- (amount first, sign after category)
const PATTERN2 = /^(\d+(?:\.\d+)?)\s*([a-zA-Z]+)?\s*([+-])(\d+(?:\.\d+)?)?$/;

// Pattern 3: rmb+100, rg-25 5.3 (category first, then sign+amount)
const PATTERN3 = /^([a-zA-Z]+)?\s*([+-])(\d+(?:\.\d+)?)\s*(\d+(?:\.\d+)?)?$/;

/**
 * Try to parse a transaction with multiple pattern attempts.
 */
export function parseTransaction(text: string): ParsedTransaction | null {
  const trimmed = text.trim();

  // Try pattern 1: +100rmb, -50sp 4.8
  let match = trimmed.match(PATTERN1);
  if (match) {
    const [, signStr, amountStr, catRaw, rateStr] = match;
    return parseMatch1(signStr, amountStr, catRaw, rateStr, trimmed);
  }

  // Try pattern 2: 100rmb+ (amount first, sign after category)
  match = trimmed.match(PATTERN2);
  if (match) {
    const [, amountStr, catRaw, signStr, rateStr] = match;
    return parseMatch2(amountStr, catRaw, signStr, rateStr, trimmed);
  }

  // Try pattern 3: rmb+100 (category first)
  match = trimmed.match(PATTERN3);
  if (match) {
    const [, catRaw, signStr, amountStr, rateStr] = match;
    return parseMatch3(catRaw, signStr, amountStr, rateStr, trimmed);
  }

  return null;
}

function parseMatch1(
  signStr: string,
  amountStr: string,
  catRaw: string | undefined,
  rateStr: string | undefined,
  raw: string
): ParsedTransaction | null {
  const inputSign: 1 | -1 = signStr === "+" ? 1 : -1;
  const amount = parseFloat(amountStr);
  const category = catRaw ? catRaw.toLowerCase() : "rmb";
  const rate = rateStr ? parseFloat(rateStr) : null;

  if (!ALL_CATEGORIES.has(category)) return null;

  return computeValue(inputSign, amount, category, rate, raw);
}

function parseMatch2(
  amountStr: string,
  catRaw: string | undefined,
  signStr: string,
  rateStr: string | undefined,
  raw: string
): ParsedTransaction | null {
  const amount = parseFloat(amountStr);
  // If there's a second number after sign, it's the rate
  const rate = rateStr ? parseFloat(rateStr) : null;

  // If category is present, use it; otherwise default to rmb
  const category = catRaw ? catRaw.toLowerCase() : "rmb";

  // For pattern like "100rmb+", the sign is at the end
  const inputSign: 1 | -1 = signStr === "+" ? 1 : -1;

  if (!ALL_CATEGORIES.has(category)) return null;

  return computeValue(inputSign, amount, category, rate, raw);
}

function parseMatch3(
  catRaw: string | undefined,
  signStr: string,
  amountStr: string,
  rateStr: string | undefined,
  raw: string
): ParsedTransaction | null {
  // Category comes first (e.g., "rmb+100" or "rg-25 5.3")
  const category = catRaw ? catRaw.toLowerCase() : "rmb";
  const inputSign: 1 | -1 = signStr === "+" ? 1 : -1;
  const amount = parseFloat(amountStr);
  const rate = rateStr ? parseFloat(rateStr) : null;

  if (!ALL_CATEGORIES.has(category)) return null;

  return computeValue(inputSign, amount, category, rate, raw);
}

function computeValue(
  inputSign: 1 | -1,
  amount: number,
  category: string,
  rate: number | null,
  raw: string
): ParsedTransaction | null {
  // RMB doesn't use rate
  if (category === "rmb") {
    if (rate !== null) return null; // RMB shouldn't have rate
    const rmbValue = inputSign * amount;
    return { inputSign, amount, category, rate: null, rmbValue, raw };
  }

  // Non-rmb requires rate
  if (rate === null || rate <= 0) return null;

  const baseValue = amount * rate;
  const isNegCat = NEGATIVE_CATEGORIES.has(category);

  // For negative categories, reverse the sign
  const effectiveSign = isNegCat ? -inputSign : inputSign;
  const rmbValue = effectiveSign * baseValue;

  return { inputSign, amount, category, rate, rmbValue, raw };
}

/**
 * Check if a message looks like it could be a bookkeeping entry.
 * Fast pre-check before full parse.
 */
export function looksLikeTransaction(text: string): boolean {
  const t = text.trim();
  // Must start with +/-, or contain category + sign, or sign after number
  return (
    (t.length >= 2 && (t[0] === "+" || t[0] === "-")) || // +100rmb
    /\d+[a-zA-Z][+-]/.test(t) || // 100rmb+, 25rg-
    (/^[a-zA-Z]+[+-]\d/.test(t)) // rmb+100, rg-25
  );
}

/**
 * Format a transaction for confirmation reply.
 */
export function formatConfirmation(tx: ParsedTransaction): string {
  const sign = tx.rmbValue >= 0 ? "+" : "";
  const rmbFormatted = `${sign}${tx.rmbValue.toFixed(2)}`;

  if (tx.category === "rmb") {
    return `✅ ${rmbFormatted}`;
  }

  const inputSign = tx.inputSign === 1 ? "+" : "-";
  return `✅ ${inputSign}${tx.amount} ${tx.category.toUpperCase()} ×${tx.rate} = ${rmbFormatted}`;
}

File: wxbot/bookkeeping-platform/bookkeeping_core/parser.py
from __future__ import annotations

import re

from .models import ParsedTransaction


NEGATIVE_CATEGORIES = {
    "rg", "sp", "gs", "xb", "it", "st", "ft", "mx", "gg", "dg", "ae",
    "lulu", "dk", "ebay", "rb", "uber", "xc", "cvs", "ymx", "psn", "chime",
    "ks", "tt", "nike", "nd", "cash", "ps", "ls", "hd",
}
POSITIVE_CATEGORIES = {"rmb"}
ALL_CATEGORIES = NEGATIVE_CATEGORIES | POSITIVE_CATEGORIES

PATTERN1 = re.compile(r"^([+-])(\d+(?:\.\d+)?)\s*([a-zA-Z]+)?\s*(\d+(?:\.\d+)?)?$")
PATTERN2 = re.compile(r"^(\d+(?:\.\d+)?)\s*([a-zA-Z]+)?\s*([+-])(\d+(?:\.\d+)?)?$")
PATTERN3 = re.compile(r"^([a-zA-Z]+)?\s*([+-])(\d+(?:\.\d+)?)\s*(\d+(?:\.\d+)?)?$")
FIXED_CODE_PATTERN = re.compile(r"^[A-Za-z0-9]{5}(?:-[A-Za-z0-9]{5}){2,}$")


def parse_transaction(text: str) -> ParsedTransaction | None:
    trimmed = text.strip()
    if not trimmed:
        return None

    match = PATTERN1.match(trimmed)
    if match:
        sign_str, amount_str, cat_raw, rate_str = match.groups()
        return _parse_match(sign_str, amount_str, cat_raw, rate_str, trimmed)

    match = PATTERN2.match(trimmed)
    if match:
        amount_str, cat_raw, sign_str, rate_str = match.groups()
        return _parse_match(sign_str, amount_str, cat_raw, rate_str, trimmed)

    match = PATTERN3.match(trimmed)
    if match:
        cat_raw, sign_str, amount_str, rate_str = match.groups()
        return _parse_match(sign_str, amount_str, cat_raw, rate_str, trimmed)

    return None


def looks_like_transaction(text: str) -> bool:
    trimmed = text.strip()
    if len(trimmed) < 2:
        return False
    first_token = trimmed.split(None, 1)[0]
    if FIXED_CODE_PATTERN.fullmatch(first_token):
        return False
    return bool(
        trimmed[0] in "+-"
        or re.search(r"\d+[a-zA-Z][+-]", trimmed)
        or re.match(r"^[a-zA-Z]+[+-]\d", trimmed)
    )


def format_confirmation(tx: ParsedTransaction) -> str:
    sign = "+" if tx.rmb_value >= 0 else "-"
    if tx.category == "rmb":
        return f"✅ {sign}{abs(tx.rmb_value):.2f}"
    input_sign = "+" if tx.input_sign > 0 else "-"
    return f"✅ {input_sign}{tx.amount:g} {tx.category.upper()} ×{tx.rate:g} = {sign}{abs(tx.rmb_value):.2f}"


def _parse_match(sign_str: str, amount_str: str, cat_raw: str | None, rate_str: str | None, raw: str) -> ParsedTransaction | None:
    input_sign = 1 if sign_str == "+" else -1
    amount = float(amount_str)
    category = (cat_raw or "rmb").lower()
    rate = float(rate_str) if rate_str else None

    if category not in ALL_CATEGORIES:
        return None

    if category == "rmb":
        if rate is not None:
            return None
        return ParsedTransaction(
            input_sign=input_sign,
            amount=amount,
            category=category,
            rate=None,
            rmb_value=input_sign * amount,
            raw=raw,
        )

    if rate is None or rate <= 0:
        return None

    effective_sign = -input_sign if category in NEGATIVE_CATEGORIES else input_sign
    return ParsedTransaction(
        input_sign=input_sign,
        amount=amount,
        category=category,
        rate=rate,
        rmb_value=effective_sign * amount * rate,
        raw=raw,
    )
